sapi/cli/php_cli.c

```diff
@@ -538,19 +538,23 @@ static void cli_register_file_handles(bool no_close) /* {{{ */
 	s_out = php_stream_open_wrapper_ex("php://stdout", "wb", 0, NULL, sc_out);
 	s_err = php_stream_open_wrapper_ex("php://stderr", "wb", 0, NULL, sc_err);
 
+	/* Release stream resources, but don't free the underlying handles. Othewrise,
+	 * extensions which write to stderr or company during mshutdown/gshutdown
+	 * won't have the expected functionality.
+	 */
+	if (no_close) {
+		s_in->flags |= PHP_STREAM_FLAG_NO_CLOSE;
+		s_out->flags |= PHP_STREAM_FLAG_NO_CLOSE;
+		s_err->flags |= PHP_STREAM_FLAG_NO_CLOSE;
+	}
+
 	if (s_in==NULL || s_out==NULL || s_err==NULL) {
 		if (s_in) php_stream_close(s_in);
 		if (s_out) php_stream_close(s_out);
 		if (s_err) php_stream_close(s_err);
 		return;
 	}
 
-	if (no_close) {
-		s_in->flags |= PHP_STREAM_FLAG_NO_CLOSE;
-		s_out->flags |= PHP_STREAM_FLAG_NO_CLOSE;
-		s_err->flags |= PHP_STREAM_FLAG_NO_CLOSE;
-	}
-
```

Comment:
Moving the code up, doesn't make much sense to me. And in this case there would even be a segfault, if any of the streams is NULL.

Reply:
While looking at this, why exactly are they all closed if _any_ is closed?

Reply:
While looking at this, why exactly are they all closed if _any_ is closed?

Reply:
The point of the PR is to not close the underlying FILE handles of stdin, stdout, and stderr. I goofed this patch for this version (the patches for 8.0, 8.1, and 8.2 are all different). That's why it should be moved up, anyway: if not, then the file handle will get closed in this edge case of one of the streams failing to open.

For master, we should consider aborting if this fails to happen, I think, rather than just not having `STDIN`, `STDOUT`, and `STDERR` be available in the script.

Reply:
> That's why it should be moved up, anyway: if not, then the file handle will get closed in this edge case of one of the streams failing to open.

Ah, I see now. :)